// lua-cli/dist/services/api.js
/**
 * Centralized API service for all Lua CLI API calls
 * This service handles all HTTP requests to various Lua endpoints
 */
// Base URLs for different environments
const BASE_URLS = {
    LOCAL: 'http://localhost:3022',
    API: 'https://api.lua.dev',
    AUTH: 'https://auth.lua.dev',
    CHAT: 'http://localhost:3001'
};
/**
 * Generic HTTP client with common error handling
 */
class HttpClient {
    async request(url, options = {}) {
        try {
            const response = await fetch(url, {
                ...options,
                headers: {
                    'Content-Type': 'application/json',
                    ...options.headers,
                },
            });
            const data = await response.json();
            if (!response.ok) {
                return {
                    success: false,
                    error: {
                        message: data.message || data.error || `HTTP ${response.status}`,
                        statusCode: response.status,
                    },
                };
            }
            return {
                success: true,
                data,
            };
        }
        catch (error) {
            return {
                success: false,
                error: {
                    message: error instanceof Error ? error.message : 'Network error',
                    statusCode: 500,
                },
            };
        }
    }
    async get(url, headers) {
        return this.request(url, { method: 'GET', headers });
    }
    async post(url, data, headers) {
        return this.request(url, {
            method: 'POST',
            body: data ? JSON.stringify(data) : undefined,
            headers,
        });
    }
    async put(url, data, headers) {
        return this.request(url, {
            method: 'PUT',
            body: data ? JSON.stringify(data) : undefined,
            headers,
        });
    }
    async delete(url, headers) {
        return this.request(url, { method: 'DELETE', headers });
    }
    async patch(url, data, headers) {
        return this.request(url, {
            method: 'PATCH',
            body: data ? JSON.stringify(data) : undefined,
            headers,
        });
    }
}
const httpClient = new HttpClient();
/**
 * Authentication API calls
 */
export class AuthApi {
    static async checkApiKey(apiKey) {
        return httpClient.get(`${BASE_URLS.LOCAL}/admin`, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async sendOtp(email) {
        return httpClient.post(`${BASE_URLS.AUTH}/otp`, { email });
    }
    static async verifyOtp(email, otp) {
        return httpClient.post(`${BASE_URLS.AUTH}/otp/verify`, { email, otp });
    }
    static async getApiKey(signInToken) {
        return httpClient.post(`${BASE_URLS.AUTH}/profile/apiKey`, undefined, {
            Authorization: `Bearer ${signInToken}`,
        });
    }
}
/**
 * Agent API calls
 */
export class AgentApi {
    static async getOrganizations(apiKey) {
        return httpClient.get(`${BASE_URLS.LOCAL}/admin`, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async getAgentTypes(apiKey) {
        return httpClient.get(`${BASE_URLS.API}/agents/self-serve/types`, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async createAgent(apiKey, agentData) {
        return httpClient.post(`${BASE_URLS.API}/agents/self-serve/create`, agentData, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async getAgent(apiKey, agentId) {
        return httpClient.get(`${BASE_URLS.API}/admin/agents/${agentId}`, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
}
/**
 * Skill API calls
 */
export class SkillApi {
    static async createSkill(apiKey, agentId, skillData) {
        return httpClient.post(`${BASE_URLS.LOCAL}/developer/skills/${agentId}`, skillData, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async pushSkill(apiKey, agentId, skillId, versionData) {
        return httpClient.post(`${BASE_URLS.LOCAL}/developer/skills/${agentId}/${skillId}/version`, versionData, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async pushDevSkill(apiKey, agentId, skillId, versionData) {
        return httpClient.post(`${BASE_URLS.LOCAL}/developer/skills/${agentId}/${skillId}/version/sandbox`, versionData, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async updateDevSkill(apiKey, agentId, skillId, sandboxVersionId, versionData) {
        return httpClient.put(`${BASE_URLS.LOCAL}/developer/skills/${agentId}/${skillId}/version/sandbox/${sandboxVersionId}`, versionData, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async getSkillVersions(apiKey, agentId, skillId) {
        return httpClient.get(`${BASE_URLS.LOCAL}/developer/skills/${agentId}/${skillId}/versions`, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async publishSkillVersion(apiKey, agentId, skillId, version) {
        return httpClient.put(`${BASE_URLS.LOCAL}/developer/skills/${agentId}/${skillId}/${version}/publish`, undefined, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
}
/**
 * Chat API calls
 */
export class ChatApi {
    static async sendMessage(agentId, chatData, apiKey) {
        return httpClient.post(`${BASE_URLS.CHAT}/chat/generate/${agentId}?channel=dev`, chatData, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
}
/**
 * Tool API calls (for compile command)
 */
export class ToolApi {
    static async getTool(url, apiKey) {
        return httpClient.get(url, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async postTool(url, data, apiKey) {
        return httpClient.post(url, data, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async putTool(url, data, apiKey) {
        return httpClient.put(url, data, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async deleteTool(url, apiKey) {
        return httpClient.delete(url, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
    static async patchTool(url, data, apiKey) {
        return httpClient.patch(url, data, {
            Authorization: `Bearer ${apiKey}`,
        });
    }
}
/**
 * Main API service that exports all API classes
 */
export const ApiService = {
    Auth: AuthApi,
    Agent: AgentApi,
    Skill: SkillApi,
    Chat: ChatApi,
    Tool: ToolApi,
};
