// lua-cli/dist/services/auth.js
import keytar from "keytar";
import { ApiService } from "./api.js";
const SERVICE = "lua-cli";
const ACCOUNT = "api-key";
export async function saveApiKey(apiKey) {
    await keytar.setPassword(SERVICE, ACCOUNT, apiKey);
}
export async function loadApiKey() {
    return keytar.getPassword(SERVICE, ACCOUNT);
}
export async function deleteApiKey() {
    return keytar.deletePassword(SERVICE, ACCOUNT);
}
export async function checkApiKey(apiKey) {
    const result = await ApiService.Auth.checkApiKey(apiKey);
    if (!result.success) {
        console.error(`❌ Invalid API key`);
        process.exit(1);
    }
    return result.data;
}
export async function requestEmailOTP(email) {
    try {
        const result = await ApiService.Auth.sendOtp(email);
        return result.success;
    }
    catch (error) {
        console.error("❌ Error requesting OTP:", error);
        return false;
    }
}
export async function verifyOTPAndGetToken(email, pin) {
    try {
        const result = await ApiService.Auth.verifyOtp(email, pin);
        return result.success ? result.data.signInToken : null;
    }
    catch (error) {
        console.error("❌ Error verifying OTP:", error);
        return null;
    }
}
export async function generateApiKey(signInToken) {
    try {
        const result = await ApiService.Auth.getApiKey(signInToken);
        return result.success ? result.data.apiKey : null;
    }
    catch (error) {
        console.error("❌ Error generating API key:", error);
        return null;
    }
}
export async function createSkill(apiKey, agentId, skillName) {
    try {
        const response = await fetch(`http://localhost:3022/developer/skills/${agentId}`, {
            method: "POST",
            headers: {
                "accept": "application/json",
                "Authorization": `Bearer ${apiKey}`,
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                name: skillName
            })
        });
        const data = await response.json();
        if (response.ok) {
            return {
                success: true,
                data: data
            };
        }
        else {
            return {
                success: false,
                error: data
            };
        }
    }
    catch (error) {
        console.error("❌ Error creating skill:", error);
        return {
            success: false,
            error: {
                message: "Network error",
                error: "Failed to connect to server",
                statusCode: 500
            }
        };
    }
}
export async function getAgentTypes(apiKey) {
    try {
        const response = await fetch('https://api.lua.dev/agents/self-serve/types', {
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${apiKey}`,
                'Content-Type': 'application/json'
            }
        });
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return await response.json();
    }
    catch (error) {
        console.error('❌ Error fetching agent types:', error);
        throw error;
    }
}
export async function createAgent(apiKey, agentData) {
    try {
        const response = await fetch('https://api.lua.dev/agents/self-serve/create', {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${apiKey}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(agentData)
        });
        const data = await response.json();
        if (response.ok) {
            return {
                success: true,
                data: data
            };
        }
        else {
            return {
                success: false,
                error: data
            };
        }
    }
    catch (error) {
        console.error('❌ Error creating agent:', error);
        return {
            success: false,
            error: {
                message: 'Network error',
                error: 'Failed to connect to server',
                statusCode: 500
            }
        };
    }
}
export async function getAgentDetails(apiKey, agentId) {
    try {
        const response = await fetch(`https://api.lua.dev/admin/agents/${agentId}`, {
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${apiKey}`,
                'Content-Type': 'application/json'
            }
        });
        const data = await response.json();
        if (response.ok) {
            return {
                success: true,
                data: data
            };
        }
        else {
            return {
                success: false,
                error: data
            };
        }
    }
    catch (error) {
        console.error('❌ Error fetching agent details:', error);
        return {
            success: false,
            error: {
                message: 'Network error',
                error: 'Failed to connect to server',
                statusCode: 500
            }
        };
    }
}
