// lua-cli/dist/commands/configure.js
import inquirer from "inquirer";
import { saveApiKey, checkApiKey, requestEmailOTP, verifyOTPAndGetToken, generateApiKey } from "../services/auth.js";
import { withErrorHandling, clearPromptLines, writeProgress, writeSuccess } from "../utils/cli.js";
export async function configureCommand() {
    return withErrorHandling(async () => {
        // Choose authentication method
        const { authMethod } = await inquirer.prompt([
            {
                type: "list",
                name: "authMethod",
                message: "Choose authentication method:",
                choices: [
                    { name: "API Key", value: "api-key" },
                    { name: "Email", value: "email" }
                ]
            }
        ]);
        // Clear the authentication method prompt lines
        clearPromptLines(2);
        if (authMethod === "api-key") {
            // Existing API key flow
            const answers = await inquirer.prompt([
                {
                    type: "password",
                    name: "apiKey",
                    message: "Enter your API key",
                    mask: "*",
                },
            ]);
            const data = await checkApiKey(answers.apiKey);
            if (!data) {
                console.error("❌ Invalid API key");
                process.exit(1);
            }
            await saveApiKey(answers.apiKey);
            writeSuccess("✅ API key saved securely.");
        }
        else if (authMethod === "email") {
            // Email authentication flow
            const { email } = await inquirer.prompt([
                {
                    type: "input",
                    name: "email",
                    message: "Enter your email address:",
                    validate: (input) => {
                        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
                        return emailRegex.test(input) || "Please enter a valid email address";
                    }
                }
            ]);
            // Clear the email prompt lines
            clearPromptLines(2);
            writeProgress("📧 Sending OTP to your email...");
            const otpSent = await requestEmailOTP(email);
            if (!otpSent) {
                console.error("❌ Failed to send OTP. Please try again.");
                process.exit(1);
            }
            writeProgress("✅ OTP sent successfully!");
            // OTP verification with retry logic
            let otpVerified = false;
            let attempts = 0;
            let signInToken = null;
            const maxAttempts = 3;
            while (!otpVerified && attempts < maxAttempts) {
                const { pin } = await inquirer.prompt([
                    {
                        type: "input",
                        name: "pin",
                        message: "Enter the OTP code:",
                        validate: (input) => {
                            return input.length === 6 || "OTP must be 6 digits";
                        }
                    }
                ]);
                // Clear the OTP prompt lines
                clearPromptLines(2);
                writeProgress("🔐 Verifying OTP...");
                signInToken = await verifyOTPAndGetToken(email, pin);
                if (signInToken) {
                    otpVerified = true;
                    writeProgress("✅ OTP verified successfully!");
                    writeProgress("🔑 Generating API key...");
                }
                else {
                    attempts++;
                    if (attempts < maxAttempts) {
                        console.log(`❌ Invalid OTP. ${maxAttempts - attempts} attempts remaining.`);
                    }
                    else {
                        console.error("❌ Invalid OTP. Maximum attempts reached. Please try again later.");
                        process.exit(1);
                    }
                }
            }
            const apiKey = await generateApiKey(signInToken);
            if (!apiKey) {
                console.error("❌ Failed to generate API key. Please try again.");
                process.exit(1);
            }
            await saveApiKey(apiKey);
            writeSuccess("✅ API key generated and saved securely.");
        }
    }, "configuration");
}
