// lua-cli/dist/commands/deploy-new.js
"use strict";
// import fs from 'fs';
// import path from 'path';
// import inquirer from 'inquirer';
// import { loadApiKey, checkApiKey } from '../services/auth.js';
// import { readSkillConfig } from '../utils/files.js';
// export interface VersionInfo {
//   version: string;
//   createdDate: string;
//   createdBy: string;
//   isCurrent: boolean;
//   createdByEmail: string;
//   createdByFullName: string;
// }
// export interface VersionsResponse {
//   versions: VersionInfo[];
// }
// export interface PublishResponse {
//   message: string;
//   skillId: string;
//   activeVersionId: string;
//   publishedAt: string;
// }
// export async function fetchVersions(apiKey: string, agentId: string, skillId: string): Promise<VersionsResponse> {
//   try {
//     const response = await fetch(`http://localhost:3022/developer/skills/${agentId}/${skillId}/versions`, {
//       method: "GET",
//       headers: {
//         "accept": "application/json",
//         "Authorization": `Bearer ${apiKey}`
//       }
//     });
//     if (!response.ok) {
//       throw new Error(`Failed to fetch versions: ${response.status} ${response.statusText}`);
//     }
//     return await response.json() as VersionsResponse;
//   } catch (error) {
//     console.error("❌ Error fetching versions:", error);
//     throw error;
//   }
// }
// export async function publishVersion(apiKey: string, agentId: string, skillId: string, version: string): Promise<PublishResponse> {
//   try {
//     const response = await fetch(`http://localhost:3022/developer/skills/${agentId}/${skillId}/${version}/publish`, {
//       method: "PUT",
//       headers: {
//         "accept": "application/json",
//         "Authorization": `Bearer ${apiKey}`
//       }
//     });
//     if (!response.ok) {
//       throw new Error(`Failed to publish version: ${response.status} ${response.statusText}`);
//     }
//     return await response.json() as PublishResponse;
//   } catch (error) {
//     console.error("❌ Error publishing version:", error);
//     throw error;
//   }
// }
// function formatVersionChoice(version: VersionInfo): string {
//   const currentIndicator = version.isCurrent ? ' (CURRENT)' : '';
//   const date = new Date(version.createdDate).toLocaleDateString();
//   return `${version.version}${currentIndicator} - Created: ${date} by ${version.createdByEmail}`;
// }
// export async function deployCommand(): Promise<void> {
//   try {
//     // Check if we're in a skill directory
//     const config = readSkillConfig();
//     if (!config || !config.agent?.agentId || !config.skill?.skillId) {
//       console.error("❌ No lua.skill.yaml found or missing agentId/skillId. Please run this command from a skill directory.");
//       process.exit(1);
//     }
//     // Load API key
//     const apiKey = await loadApiKey();
//     if (!apiKey) {
//       console.error("❌ No API key found. Please run 'lua configure' to set up your API key.");
//       process.exit(1);
//     }
//     // Validate API key
//     const userData = await checkApiKey(apiKey);
//     console.log(`✅ Authenticated as ${userData.email}`);
//     // Fetch available versions
//     console.log("🔄 Fetching available versions...");
//     const versionsResponse = await fetchVersions(apiKey, config.agent.agentId, config.skill.skillId);
//     if (!versionsResponse.versions || versionsResponse.versions.length === 0) {
//       console.log("❌ No versions found for this skill.");
//       process.exit(1);
//     }
//     // Sort versions by creation date (newest first)
//     const sortedVersions = versionsResponse.versions.sort((a, b) => 
//       new Date(b.createdDate).getTime() - new Date(a.createdDate).getTime()
//     );
//     // Let user select a version
//     const { selectedVersion } = await inquirer.prompt([
//       {
//         type: "list",
//         name: "selectedVersion",
//         message: "Select a version to deploy:",
//         choices: sortedVersions.map(version => ({
//           name: formatVersionChoice(version),
//           value: version.version
//         }))
//       }
//     ]);
//     // Show warning and confirm deployment
//     const { confirmed } = await inquirer.prompt([
//       {
//         type: "confirm",
//         name: "confirmed",
//         message: "⚠️  Warning: This version will be deployed to all users. Do you want to proceed?",
//         default: false
//       }
//     ]);
//     if (!confirmed) {
//       console.log("❌ Deployment cancelled.");
//       process.exit(0);
//     }
//     // Publish the selected version
//     console.log(`🔄 Publishing version ${selectedVersion}...`);
//     const publishResponse = await publishVersion(apiKey, config.agent.agentId, config.skill.skillId, selectedVersion);
//     console.log("✅ Deployment successful!");
//     console.log(`📦 Version: ${publishResponse.activeVersionId}`);
//     console.log(`🆔 Skill ID: ${publishResponse.skillId}`);
//     console.log(`📅 Published at: ${new Date(publishResponse.publishedAt).toLocaleString()}`);
//     console.log(`💬 ${publishResponse.message}`);
//   } catch (error: any) {
//     console.error("❌ Error during deployment:", error.message);
//     process.exit(1);
//   }
// }
