// lua-cli/dist/commands/deploy.js
import inquirer from 'inquirer';
import { loadApiKey, checkApiKey } from '../services/auth.js';
import { ApiService } from '../services/api.js';
import { readSkillConfig } from '../utils/files.js';
import { withErrorHandling, clearPromptLines, writeProgress, writeSuccess } from '../utils/cli.js';
export async function fetchVersions(apiKey, agentId, skillId) {
    try {
        const response = await ApiService.Skill.getSkillVersions(apiKey, agentId, skillId);
        if (!response.success) {
            throw new Error(`Failed to fetch versions: ${response.error?.message || 'Unknown error'}`);
        }
        return response.data;
    }
    catch (error) {
        console.error("❌ Error fetching versions:", error);
        throw error;
    }
}
export async function publishVersion(apiKey, agentId, skillId, version) {
    try {
        const response = await ApiService.Skill.publishSkillVersion(apiKey, agentId, skillId, version);
        if (!response.success) {
            throw new Error(`Failed to publish version: ${response.error?.message || 'Unknown error'}`);
        }
        return response.data;
    }
    catch (error) {
        console.error("❌ Error publishing version:", error);
        throw error;
    }
}
function formatVersionChoice(version) {
    const currentIndicator = version.isCurrent ? ' (CURRENT)' : '';
    const date = new Date(version.createdDate).toLocaleDateString();
    return `${version.version}${currentIndicator} - Created: ${date} by ${version.createdByEmail}`;
}
export async function deployCommand() {
    return withErrorHandling(async () => {
        // Check if we're in a skill directory
        const config = readSkillConfig();
        if (!config || !config.agent?.agentId || !config.skill?.skillId) {
            console.error("❌ No lua.skill.yaml found or missing agentId/skillId. Please run this command from a skill directory.");
            process.exit(1);
        }
        // Load API key
        const apiKey = await loadApiKey();
        if (!apiKey) {
            console.error("❌ No API key found. Please run 'lua configure' to set up your API key.");
            process.exit(1);
        }
        // Validate API key
        const userData = await checkApiKey(apiKey);
        writeProgress("✅ Authenticated");
        // Fetch available versions
        writeProgress("🔄 Fetching available versions...");
        const versionsResponse = await fetchVersions(apiKey, config.agent.agentId, config.skill.skillId);
        if (!versionsResponse.versions || versionsResponse.versions.length === 0) {
            console.log("❌ No versions found for this skill.");
            process.exit(1);
        }
        // Sort versions by creation date (newest first)
        const sortedVersions = versionsResponse.versions.sort((a, b) => new Date(b.createdDate).getTime() - new Date(a.createdDate).getTime());
        // Let user select a version
        const { selectedVersion } = await inquirer.prompt([
            {
                type: "list",
                name: "selectedVersion",
                message: "Select a version to deploy:",
                choices: sortedVersions.map(version => ({
                    name: formatVersionChoice(version),
                    value: version.version
                }))
            }
        ]);
        // Clear the selection prompt lines
        clearPromptLines(2);
        // Show warning and confirm deployment
        const { confirmed } = await inquirer.prompt([
            {
                type: "confirm",
                name: "confirmed",
                message: "⚠️  Warning: This version will be deployed to all users. Do you want to proceed?",
                default: false
            }
        ]);
        // Clear the confirmation prompt lines
        clearPromptLines(2);
        if (!confirmed) {
            console.log("❌ Deployment cancelled.");
            process.exit(0);
        }
        // Publish the selected version
        writeProgress("🔄 Publishing version...");
        const publishResponse = await publishVersion(apiKey, config.agent.agentId, config.skill.skillId, selectedVersion);
        writeSuccess(`✅ Version ${publishResponse.activeVersionId} deployed successfully`);
    }, "deployment");
}
