// lua-cli/dist/commands/init.js
import inquirer from "inquirer";
import { fileURLToPath } from "url";
import path from "path";
import { loadApiKey, checkApiKey, createSkill, getAgentTypes, createAgent, getAgentDetails } from "../services/auth.js";
import { copyTemplateFiles, createSkillYaml } from "../utils/files.js";
import { withErrorHandling, clearPromptLines, writeProgress, writeSuccess } from "../utils/cli.js";
function getTemplateDir() {
    const __filename = fileURLToPath(import.meta.url);
    const __dirname = path.dirname(__filename);
    return path.join(__dirname, "..", "..", "template");
}
export async function initCommand() {
    return withErrorHandling(async () => {
        const apiKey = await loadApiKey();
        if (!apiKey) {
            console.error("❌ No API key found. Run `lua configure` first.");
            process.exit(1);
        }
        // Ask user if they want to create a new agent or select existing one
        const { agentChoice } = await inquirer.prompt([
            {
                type: "list",
                name: "agentChoice",
                message: "What would you like to do?",
                choices: [
                    { name: "Select existing agent", value: "existing" },
                    { name: "Create new agent", value: "create" }
                ]
            }
        ]);
        // Clear the agent choice prompt lines
        clearPromptLines(2);
        let selectedAgent;
        let selectedOrg;
        let persona;
        let welcomeMessage;
        if (agentChoice === "existing") {
            // Get user data from API only when selecting existing agent
            const userData = await checkApiKey(apiKey);
            writeProgress("✅ Authenticated");
            // Extract organizations and create choices for selection
            const orgs = userData.admin.orgs;
            if (!orgs || orgs.length === 0) {
                console.error("❌ No organizations found.");
                process.exit(1);
            }
            const orgChoices = orgs.map((org) => ({
                name: org.registeredName || org.name || 'Unknown Organization',
                value: org
            }));
            // Select organization
            const { selectedOrg: existingOrg } = await inquirer.prompt([
                {
                    type: "list",
                    name: "selectedOrg",
                    message: "Select an organization:",
                    choices: orgChoices
                }
            ]);
            selectedOrg = existingOrg;
            // Clear the organization selection prompt lines
            clearPromptLines(2);
            // Extract agents from selected organization
            if (!selectedOrg.agents || selectedOrg.agents.length === 0) {
                console.error("❌ No agents found in the selected organization.");
                process.exit(1);
            }
            const agentChoices = selectedOrg.agents.map((agent) => ({
                name: agent.name,
                value: agent
            }));
            // Select agent
            const { selectedAgent: existingAgent } = await inquirer.prompt([
                {
                    type: "list",
                    name: "selectedAgent",
                    message: "Select an agent:",
                    choices: agentChoices
                }
            ]);
            selectedAgent = existingAgent;
            // Clear the agent selection prompt lines
            clearPromptLines(2);
        }
        else {
            // Create new agent flow
            writeProgress("🔄 Fetching agent types...");
            const agentTypes = await getAgentTypes(apiKey);
            writeProgress("✅ Agent types loaded");
            // Select agent type
            const agentTypeChoices = agentTypes.map((type) => ({
                name: type.name,
                value: type
            }));
            const { selectedAgentType } = await inquirer.prompt([
                {
                    type: "list",
                    name: "selectedAgentType",
                    message: "Select an agent type:",
                    choices: agentTypeChoices
                }
            ]);
            // Clear the agent type selection prompt lines
            clearPromptLines(2);
            // Collect required metadata
            const metadata = {};
            if (selectedAgentType.requiredSubAgentMetadata && selectedAgentType.requiredSubAgentMetadata.length > 0) {
                for (const field of selectedAgentType.requiredSubAgentMetadata) {
                    const { [field]: value } = await inquirer.prompt([
                        {
                            type: "input",
                            name: field,
                            message: `Enter ${field}:`,
                            validate: (input) => input.trim().length > 0 || `${field} is required`
                        }
                    ]);
                    metadata[field] = value;
                    // Clear the metadata prompt lines
                    clearPromptLines(2);
                }
            }
            // Configure features
            const features = {};
            const availableFeatures = Object.keys(selectedAgentType.features);
            if (availableFeatures.length > 0) {
                for (const feature of availableFeatures) {
                    const { [feature]: enabled } = await inquirer.prompt([
                        {
                            type: "confirm",
                            name: feature,
                            message: `Enable ${feature} feature?`,
                            default: selectedAgentType.features[feature]?.active || false
                        }
                    ]);
                    features[feature] = enabled;
                    // Clear the feature prompt lines
                    clearPromptLines(2);
                }
            }
            // Collect persona information
            const businessTypes = [
                "Retail & Consumer Goods (e.g. clothing store, bookstore)",
                "Food & Beverage (e.g. restaurant, bakery)",
                "Hospitality (e.g. hotel, resort)",
                "Personal services (e.g. hair salon, spa)",
                "Education (e.g. language school, online courses)",
                "Health (e.g. gym, yoga studio, clinic)"
            ];
            const brandPersonalities = [
                "Energetic & fun",
                "Refined & elegant",
                "Chatty",
                "Casual",
                "Funny",
                "Friendly"
            ];
            const { businessName } = await inquirer.prompt([
                {
                    type: "input",
                    name: "businessName",
                    message: "Enter business name:",
                    default: "Private"
                }
            ]);
            clearPromptLines(2);
            const { agentName } = await inquirer.prompt([
                {
                    type: "input",
                    name: "agentName",
                    message: "Enter agent name:",
                    validate: (input) => input.trim().length > 0 || "Agent name is required"
                }
            ]);
            clearPromptLines(2);
            const { businessType } = await inquirer.prompt([
                {
                    type: "list",
                    name: "businessType",
                    message: "Select business type:",
                    choices: businessTypes
                }
            ]);
            clearPromptLines(2);
            const { brandPersonality } = await inquirer.prompt([
                {
                    type: "list",
                    name: "brandPersonality",
                    message: "Select brand personality:",
                    choices: brandPersonalities
                }
            ]);
            clearPromptLines(2);
            const { brandTraits } = await inquirer.prompt([
                {
                    type: "input",
                    name: "brandTraits",
                    message: "Enter brand traits (any traits the agent might have):",
                    default: ""
                }
            ]);
            clearPromptLines(2);
            // For new agents, we need to handle organization selection
            // For now, we'll create a new organization with the business name
            // In the future, we could add organization selection here too
            // Create agent
            writeProgress("🔄 Creating agent...");
            const agentData = {
                id: `agent_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
                type: selectedAgentType.id,
                metadata,
                persona: {
                    agentName,
                    businessType,
                    brandPersonality,
                    brandTraits
                },
                features,
                channels: [],
                org: {
                    registeredName: businessName
                }
            };
            const createAgentResult = await createAgent(apiKey, agentData);
            if (!createAgentResult.success || !createAgentResult.data) {
                console.error("❌ Failed to create agent:", createAgentResult.error?.message || "Unknown error");
                process.exit(1);
            }
            writeProgress("✅ Agent created successfully!");
            // Wait 30 seconds and get agent details
            writeProgress("⏳ Waiting for agent to be ready...");
            await new Promise(resolve => setTimeout(resolve, 30000));
            writeProgress("🔄 Fetching agent details...");
            const agentDetailsResult = await getAgentDetails(apiKey, createAgentResult.data.agentId);
            if (!agentDetailsResult.success || !agentDetailsResult.data) {
                console.error("❌ Failed to fetch agent details:", agentDetailsResult.error?.message || "Unknown error");
                process.exit(1);
            }
            selectedAgent = {
                agentId: createAgentResult.data.agentId,
                name: createAgentResult.data.name
            };
            persona = agentDetailsResult.data.persona;
            welcomeMessage = agentDetailsResult.data.welcomeMessage;
            // For new agents, get the organization ID from the created agent response
            selectedOrg = {
                id: createAgentResult.data.org.id
            };
            writeProgress("✅ Agent details loaded");
        }
        // Get skill name and handle retries
        let skillName = "";
        let skillData = null;
        let retry = true;
        while (retry) {
            const { skillName: inputSkillName } = await inquirer.prompt([
                {
                    type: "input",
                    name: "skillName",
                    message: "Enter a name for your skill:",
                    default: "My Lua Skill"
                }
            ]);
            skillName = inputSkillName;
            // Create skill via API
            writeProgress("🔄 Creating skill...");
            const result = await createSkill(apiKey, selectedAgent.agentId, skillName);
            if (result.success && result.data) {
                skillData = result.data;
                writeProgress("✅ Skill created successfully!");
                retry = false;
            }
            else if (result.error) {
                console.error(`❌ ${result.error.message}`);
                const { shouldRetry } = await inquirer.prompt([
                    {
                        type: "confirm",
                        name: "shouldRetry",
                        message: "Would you like to try with a different name?",
                        default: true
                    }
                ]);
                // Clear the retry prompt lines
                clearPromptLines(2);
                if (!shouldRetry) {
                    writeProgress("❌ Skill creation cancelled.");
                    process.exit(1);
                }
            }
            else {
                console.error("❌ Failed to create skill. Please try again.");
                process.exit(1);
            }
        }
        // Create lua.skill.yaml file with skill ID, actual skill name, persona and welcome message from backend
        const actualSkillName = skillData?.name || skillName;
        // Copy template files
        const templateDir = getTemplateDir();
        const currentDir = process.cwd();
        copyTemplateFiles(templateDir, currentDir);
        createSkillYaml(selectedAgent.agentId, selectedOrg.id, actualSkillName, skillData?.id, persona, welcomeMessage);
        writeProgress("✅ Created lua.skill.yaml");
        writeProgress("✅ Copied template files");
        writeSuccess("✅ Lua skill project initialized successfully!");
    }, "initialization");
}
