// lua-cli/dist/commands/push.js
import fs from 'fs';
import path from 'path';
import inquirer from 'inquirer';
import { compileCommand } from './compile.js';
import { checkApiKey, loadApiKey } from '../services/auth.js';
import { ApiService } from '../services/api.js';
import { readSkillConfig } from '../utils/files.js';
import { withErrorHandling, clearPromptLines, writeProgress, writeSuccess } from '../utils/cli.js';
export async function pushVersion(apiKey, agentId, skillId, versionData) {
    try {
        const response = await ApiService.Skill.pushSkill(apiKey, agentId, skillId, versionData);
        if (response.success) {
            return {
                success: true,
                data: response.data
            };
        }
        else {
            return {
                success: false,
                error: {
                    message: response.error?.message || 'Unknown error',
                    error: response.error?.message || 'Unknown error',
                    statusCode: response.error?.statusCode || 500
                }
            };
        }
    }
    catch (error) {
        console.error("❌ Error pushing version:", error);
        return {
            success: false,
            error: {
                message: "Network error",
                error: "Failed to connect to server",
                statusCode: 500
            }
        };
    }
}
function readConfigVersion() {
    const config = readSkillConfig();
    return config?.skill?.version || null;
}
function readDeployJson() {
    const deployPath = path.join(process.cwd(), '.lua', 'deploy.json');
    if (!fs.existsSync(deployPath)) {
        return null;
    }
    const deployContent = fs.readFileSync(deployPath, 'utf8');
    return JSON.parse(deployContent);
}
export async function pushCommand() {
    return withErrorHandling(async () => {
        // Check if we're in a skill directory
        const config = readSkillConfig();
        if (!config) {
            console.error("❌ No lua.skill.yaml found. Please run this command from a skill directory.");
            process.exit(1);
        }
        // Read version from config
        const version = config.skill?.version;
        if (!version) {
            console.error("❌ No version found in skill configuration");
            process.exit(1);
        }
        // Confirm with user
        const { confirmed } = await inquirer.prompt([
            {
                type: "confirm",
                name: "confirmed",
                message: `Are you sure you want to push version ${version}?`,
                default: false
            }
        ]);
        // Clear the confirmation prompt lines
        clearPromptLines(2);
        if (!confirmed) {
            console.log("❌ Push cancelled.");
            process.exit(0);
        }
        // Load API key
        const apiKey = await loadApiKey();
        if (!apiKey) {
            console.error("❌ No API key found. Please run 'lua configure' to set up your API key.");
            process.exit(1);
        }
        // Validate API key
        const userData = await checkApiKey(apiKey);
        writeProgress("✅ Authenticated");
        // Compile the skill first
        writeProgress("🔄 Compiling skill...");
        await compileCommand();
        // Read deploy.json
        const deployData = readDeployJson();
        if (!deployData) {
            console.error("❌ No deploy.json found. Compilation may have failed.");
            process.exit(1);
        }
        // Verify version matches
        if (deployData.version !== version) {
            console.error(`❌ Version mismatch: config has ${version}, deploy.json has ${deployData.version}`);
            process.exit(1);
        }
        // Extract agentId and skillId from config
        const agentId = config.agent?.agentId;
        const skillId = config.skill?.skillId;
        if (!agentId || !skillId) {
            console.error("❌ Missing agentId or skillId in skill configuration");
            process.exit(1);
        }
        // Push version
        writeProgress("🔄 Pushing version to server...");
        const result = await pushVersion(apiKey, agentId, skillId, deployData);
        if (result.success && result.data) {
            writeSuccess(`✅ Version ${result.data.version} pushed successfully`);
        }
        else if (result.error) {
            console.error(`❌ ${result.error.message}`);
            process.exit(1);
        }
        else {
            console.error("❌ Failed to push version. Please try again.");
            process.exit(1);
        }
    }, "push");
}
