// lua-cli/dist/commands/test.js
import fs from "fs";
import path from "path";
import inquirer from "inquirer";
import { compileCommand } from "./compile.js";
import { gunzipSync } from "zlib";
import { Buffer } from "buffer";
import { createRequire } from "module";
import vm from "vm";
import { withErrorHandling, clearPromptLines, writeProgress, writeSuccess } from "../utils/cli.js";
// Decompression utility
function decompressCode(compressedCode) {
    const buffer = Buffer.from(compressedCode, 'base64');
    return gunzipSync(buffer).toString('utf8');
}
export async function testCommand() {
    return withErrorHandling(async () => {
        writeProgress("🧪 Testing Lua skill...");
        // First, compile the code
        writeProgress("📦 Compiling code first...");
        await compileCommand();
        // Check if .lua directory exists
        const luaDir = path.join(process.cwd(), ".lua");
        if (!fs.existsSync(luaDir)) {
            throw new Error(".lua directory not found. Run 'lua compile' first.");
        }
        // Read deploy.json
        const deployJsonPath = path.join(luaDir, "deploy.json");
        if (!fs.existsSync(deployJsonPath)) {
            throw new Error("deploy.json not found. Run 'lua compile' first.");
        }
        const deployData = JSON.parse(fs.readFileSync(deployJsonPath, "utf8"));
        if (!deployData.tools || deployData.tools.length === 0) {
            throw new Error("No tools found in deploy.json");
        }
        // Let user select a tool using picker
        const toolChoices = deployData.tools.map((tool) => ({
            name: `${tool.name} - ${tool.description}`,
            value: tool
        }));
        const { selectedTool } = await inquirer.prompt([
            {
                type: 'list',
                name: 'selectedTool',
                message: '🔧 Select a tool to test:',
                choices: toolChoices,
                pageSize: 10
            }
        ]);
        // Clear the tool selection prompt lines
        clearPromptLines(2);
        writeProgress(`✅ Selected tool: ${selectedTool.name}`);
        // Get input values for each required parameter using inquirer
        const inputValues = {};
        const inputSchema = selectedTool.inputSchema;
        if (inputSchema.properties) {
            writeProgress("\n📝 Enter input values:");
            const inputPrompts = [];
            for (const [key, value] of Object.entries(inputSchema.properties)) {
                const property = value;
                const isRequired = inputSchema.required?.includes(key) || false;
                let promptType = 'input';
                let validate = undefined;
                // Set up validation and input type based on schema
                switch (property.type) {
                    case "string":
                        if (isRequired) {
                            validate = (input) => input.trim() !== "" || `${key} is required`;
                        }
                        break;
                    case "number":
                        promptType = 'number';
                        if (isRequired) {
                            validate = (input) => !isNaN(input) || `${key} must be a valid number`;
                        }
                        break;
                    case "boolean":
                        promptType = 'confirm';
                        break;
                    default:
                        if (isRequired) {
                            validate = (input) => input.trim() !== "" || `${key} is required`;
                        }
                }
                inputPrompts.push({
                    type: promptType,
                    name: key,
                    message: `${key}${isRequired ? " (required)" : " (optional)"}:`,
                    validate: validate,
                    when: isRequired ? true : (answers) => {
                        // For optional fields, ask if user wants to provide a value
                        return true;
                    }
                });
            }
            const answers = await inquirer.prompt(inputPrompts);
            // Convert answers to proper types
            for (const [key, value] of Object.entries(answers)) {
                const property = inputSchema.properties[key];
                switch (property.type) {
                    case "string":
                        inputValues[key] = value;
                        break;
                    case "number":
                        inputValues[key] = parseFloat(value);
                        break;
                    case "boolean":
                        inputValues[key] = value;
                        break;
                    default:
                        inputValues[key] = value;
                }
            }
        }
        writeProgress("\n🚀 Executing tool...");
        writeProgress(`Input: ${JSON.stringify(inputValues, null, 2)}`);
        // Get the execute function string directly from the selected tool and decompress it
        const toolCode = decompressCode(selectedTool.execute);
        // Execute the tool
        try {
            // Create a CommonJS context for execution
            const require = createRequire(process.cwd() + '/package.json');
            // Create a sandbox context with require and other necessary globals
            const sandbox = {
                require,
                console,
                Buffer,
                setTimeout,
                setInterval,
                clearTimeout,
                clearInterval,
                process,
                global: globalThis,
                __dirname: process.cwd(),
                __filename: path.join(process.cwd(), 'index.ts'),
                module: { exports: {} },
                exports: {},
                // Web APIs
                fetch: globalThis.fetch,
                URLSearchParams: globalThis.URLSearchParams,
                URL: globalThis.URL,
                Headers: globalThis.Headers,
                Request: globalThis.Request,
                Response: globalThis.Response,
                // Additional Node.js globals that might be needed
                Object,
                Array,
                String,
                Number,
                Boolean,
                Date,
                Math,
                JSON,
                Error,
                TypeError,
                ReferenceError,
                SyntaxError,
                // Node.js specific globals
                globalThis,
                // Additional globals that might be referenced
                undefined: undefined,
                null: null,
                Infinity: Infinity,
                NaN: NaN
            };
            // Create the CommonJS wrapper code
            const commonJsWrapper = `
const executeFunction = ${toolCode};

// Export the function for testing
module.exports = async (input) => {
  return await executeFunction(input);
};
`;
            // Execute the code in the sandbox
            const context = vm.createContext(sandbox);
            vm.runInContext(commonJsWrapper, context);
            // Get the exported function
            const executeFunction = context.module.exports;
            const result = await executeFunction(inputValues);
            writeSuccess("✅ Tool execution successful!");
            console.log(`Output: ${JSON.stringify(result, null, 2)}`);
        }
        catch (executionError) {
            console.error("\n❌ Tool execution failed:");
            console.error(executionError.message);
            if (executionError.stack) {
                console.error(executionError.stack);
            }
        }
    }, "testing");
}
