// lua-cli/dist/utils/cli.js
/**
 * Centralized CLI utilities for consistent error handling and output management
 */
/**
 * Wraps a command function with standardized error handling
 * Handles SIGINT (Ctrl+C) gracefully and provides consistent error messages
 */
export async function withErrorHandling(commandFn, commandName) {
    try {
        return await commandFn();
    }
    catch (error) {
        if (error.name === 'ExitPromptError') {
            console.log("\n❌ Operation cancelled.");
            process.exit(0);
        }
        console.error(`❌ Error during ${commandName}:`, error.message);
        process.exit(1);
    }
}
/**
 * Clears the specified number of lines from the terminal
 * Used to clean up inquirer prompt output ONLY
 * Should NOT be used to clear user input commands
 */
export function clearPromptLines(count = 1) {
    for (let i = 0; i < count; i++) {
        process.stdout.write('\x1b[1A\x1b[2K'); // Move up 1 line and clear it
    }
}
/**
 * Writes a progress message that will be replaced by the next message
 * Uses carriage return to overwrite the current line
 */
export function writeProgress(message) {
    process.stdout.write(`${message}\r`);
}
/**
 * Writes a final success message that will remain visible
 * Uses console.log to ensure it stays in the output
 */
export function writeSuccess(message) {
    console.log(message);
}
/**
 * Writes an error message
 * Uses console.error for proper error output
 */
export function writeError(message) {
    console.error(message);
}
/**
 * Writes an info message
 * Uses console.log for informational output
 */
export function writeInfo(message) {
    console.log(message);
}
