// lua-cli/dist/utils/files.js
import * as fs from "fs";
import * as path from "path";
import pkg from 'js-yaml';
const { load } = pkg;
export function copyTemplateFiles(templateDir, targetDir) {
    const files = fs.readdirSync(templateDir);
    for (const file of files) {
        // Skip node_modules and package-lock.json to avoid circular dependencies
        if (file === 'node_modules' || file === 'package-lock.json') {
            continue;
        }
        const srcPath = path.join(templateDir, file);
        const destPath = path.join(targetDir, file);
        if (fs.statSync(srcPath).isDirectory()) {
            fs.mkdirSync(destPath, { recursive: true });
            copyTemplateFiles(srcPath, destPath);
        }
        else if (file === 'package.json') {
            // Special handling for package.json to update lua-cli version
            updatePackageJson(srcPath, destPath);
        }
        else {
            fs.copyFileSync(srcPath, destPath);
        }
    }
}
function updatePackageJson(srcPath, destPath) {
    // Read the template package.json
    const templatePackageJson = JSON.parse(fs.readFileSync(srcPath, 'utf8'));
    // Get the current CLI version from the main package.json
    const mainPackageJsonPath = path.join(process.cwd(), 'package.json');
    const mainPackageJson = JSON.parse(fs.readFileSync(mainPackageJsonPath, 'utf8'));
    const currentCliVersion = mainPackageJson.version;
    // Update the lua-cli dependency version
    if (templatePackageJson.dependencies && templatePackageJson.dependencies['lua-cli']) {
        templatePackageJson.dependencies['lua-cli'] = `^${currentCliVersion}`;
    }
    // Write the updated package.json
    fs.writeFileSync(destPath, JSON.stringify(templatePackageJson, null, 2) + '\n');
}
export function createSkillYaml(agentId, orgId, skillName, skillId, persona, welcomeMessage) {
    const skillIdSection = skillId ? `  skillId: "${skillId}"\n` : '';
    // Handle multiline strings properly for YAML
    const personaSection = persona ? `  persona: |\n${persona.split('\n').map(line => `    ${line}`).join('\n')}\n` : '';
    const welcomeMessageSection = welcomeMessage ? `  welcomeMessage: "${welcomeMessage}"\n` : '';
    const yamlContent = `agent:
  agentId: "${agentId}"
  orgId: "${orgId}"
${personaSection}${welcomeMessageSection}
skill:
  name: "${skillName}"
  version: "0.0.1"
${skillIdSection}`;
    fs.writeFileSync("lua.skill.yaml", yamlContent);
}
export function readSkillYaml() {
    const yamlPath = path.join(process.cwd(), 'lua.skill.yaml');
    if (!fs.existsSync(yamlPath)) {
        return null;
    }
    const yamlContent = fs.readFileSync(yamlPath, 'utf8');
    return load(yamlContent);
}
export function readSkillConfig() {
    // Read YAML config file
    return readSkillYaml();
}
export function updateSkillYamlPersona(persona, welcomeMessage) {
    const yamlPath = path.join(process.cwd(), 'lua.skill.yaml');
    if (!fs.existsSync(yamlPath)) {
        console.error('❌ No lua.skill.yaml found. Please run this command from a skill directory.');
        return;
    }
    const yamlContent = fs.readFileSync(yamlPath, 'utf8');
    const data = load(yamlContent);
    // Update persona
    if (!data.agent) {
        data.agent = {};
    }
    data.agent.persona = persona;
    // Update welcome message if provided
    if (welcomeMessage) {
        data.agent.welcomeMessage = welcomeMessage;
    }
    // Convert back to YAML with proper formatting
    const updatedYamlContent = `agent:
  agentId: "${data.agent.agentId}"
  orgId: "${data.agent.orgId}"
  persona: |
${persona.split('\n').map(line => `    ${line}`).join('\n')}
  welcomeMessage: "${data.agent.welcomeMessage || ''}"

skill:
  name: "${data.skill.name}"
  version: "${data.skill.version}"
  skillId: "${data.skill.skillId || ''}"
`;
    fs.writeFileSync(yamlPath, updatedYamlContent);
}
