// lua-cli/dist/skill.js
import { assertValidToolName } from "./types/index.js";
export class LuaSkill {
    /**
     * Creates a new LuaSkill instance
     *
     * @param config - Configuration object containing skill metadata
     * @param config.description - Short description of what the skill does (1-2 sentences)
     * @param config.context - Detailed explanation of how the agent should use the tools
     *
     * @example
     * ```typescript
     * const skill = new LuaSkill({
     *   description: "Weather and calculator utilities",
     *   context: "This skill provides weather information for any city and basic mathematical operations. Use get_weather for current conditions and calculator for arithmetic operations."
     * });
     * ```
     */
    constructor(config) {
        this.tools = [];
        this.description = config.description;
        this.context = config.context;
    }
    addTool(tool) {
        // Validate the tool name before adding it
        assertValidToolName(tool.name);
        this.tools.push(tool);
    }
    async run(input) {
        const tool = this.tools.find(tool => tool.name === input.tool);
        if (!tool) {
            throw new Error(`Tool ${input.tool} not found`);
        }
        // Validate input against the tool's schema
        const validatedInput = tool.inputSchema.parse(input);
        return tool.execute(validatedInput);
    }
}
