// lua-cli/dist/user-data-api.js
import { readSkillConfig } from './utils/files.js';
import { loadApiKey } from './services/auth.js';
/**
 * User Data API for Lua CLI projects
 * Provides methods to interact with user data stored in the Lua system
 */
export class UserDataAPI {
    constructor(apiKey, agentId) {
        this.apiKey = apiKey || '';
        this.agentId = agentId || '';
    }
    /**
     * Initialize the API with credentials
     * If not provided, will attempt to load from stored configuration
     */
    async initialize() {
        if (!this.apiKey) {
            this.apiKey = await loadApiKey() || '';
        }
        if (!this.agentId) {
            const config = await readSkillConfig();
            this.agentId = config.agent.agentId;
        }
        if (!this.apiKey || !this.agentId) {
            throw new Error('API key and agent ID are required. Please run "lua auth configure" first.');
        }
    }
    /**
     * Get user data for the current agent
     * @returns Promise<UserDataResponse>
     */
    async get() {
        await this.initialize();
        const response = await fetch(`http://localhost:3022/developer/user/data/agent/${this.agentId}`, {
            method: 'GET',
            headers: {
                'accept': '*/*',
                'Authorization': `Bearer ${this.apiKey}`
            }
        });
        if (!response.ok) {
            // Try alternative authentication method if Bearer fails
            if (response.status === 401) {
                const altResponse = await fetch(`http://localhost:3022/developer/user/data/agent/${this.agentId}`, {
                    method: 'GET',
                    headers: {
                        'accept': '*/*',
                        'Authorization': this.apiKey
                    }
                });
                if (altResponse.ok) {
                    return await altResponse.json();
                }
            }
            throw new Error(`Failed to get user data: ${response.status} ${response.statusText}`);
        }
        return await response.json();
    }
    /**
     * Create or update user data for the current agent
     * @param data - The user data to store
     * @returns Promise<UserDataResponse>
     */
    async create(data) {
        await this.initialize();
        const response = await fetch(`http://localhost:3022/developer/user/data/agent/${this.agentId}`, {
            method: 'PUT',
            headers: {
                'accept': '*/*',
                'Authorization': `Bearer ${this.apiKey}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(data)
        });
        if (!response.ok) {
            // Try alternative authentication method if Bearer fails
            if (response.status === 401) {
                const altResponse = await fetch(`http://localhost:3022/developer/user/data/agent/${this.agentId}`, {
                    method: 'PUT',
                    headers: {
                        'accept': '*/*',
                        'Authorization': this.apiKey,
                        'Content-Type': 'application/json'
                    },
                    body: JSON.stringify(data)
                });
                if (altResponse.ok) {
                    return await altResponse.json();
                }
            }
            throw new Error(`Failed to create user data: ${response.status} ${response.statusText}`);
        }
        return await response.json();
    }
    /**
     * Update existing user data for the current agent
     * @param data - The user data to update
     * @returns Promise<UserDataResponse>
     */
    async update(data) {
        // Update is the same as create for this API
        return this.create(data);
    }
    /**
     * Clear all user data for the current agent
     * @returns Promise<{success: boolean}>
     */
    async clear() {
        await this.initialize();
        const response = await fetch(`http://localhost:3022/developer/user/data/agent/${this.agentId}`, {
            method: 'DELETE',
            headers: {
                'accept': '*/*',
                'Authorization': `Bearer ${this.apiKey}`
            }
        });
        if (!response.ok) {
            // Try alternative authentication method if Bearer fails
            if (response.status === 401) {
                const altResponse = await fetch(`http://localhost:3022/developer/user/data/agent/${this.agentId}`, {
                    method: 'DELETE',
                    headers: {
                        'accept': '*/*',
                        'Authorization': this.apiKey
                    }
                });
                if (altResponse.ok) {
                    return await altResponse.json();
                }
            }
            throw new Error(`Failed to clear user data: ${response.status} ${response.statusText}`);
        }
        return await response.json();
    }
}
/**
 * Convenience object for user data operations
 * Usage: user.data.get(), user.data.create(data), user.data.update(data), user.data.clear()
 */
export const user = {
    data: new UserDataAPI()
};
/**
 * Create a new UserDataAPI instance with custom credentials
 * @param apiKey - Optional API key
 * @param agentId - Optional agent ID
 * @returns UserDataAPI instance
 */
export function createUserDataAPI(apiKey, agentId) {
    return new UserDataAPI(apiKey, agentId);
}
