// lua-cli/template/tools/UserPreferencesTool.ts
import { LuaTool } from 'lua-cli/skill';
import { user } from 'lua-cli/user-data-api';
import { z } from 'zod';

/**
 * Example tool demonstrating User Data API usage
 * This tool allows users to manage their preferences
 */
export class UserPreferencesTool extends LuaTool {
  constructor() {
    super('user-preferences', 'Manage user preferences and settings');
  }

  async execute(input: { action: string; key?: string; value?: string }) {
    const { action, key, value } = input;

    try {
      switch (action) {
        case 'get':
          const userData = await user.data.get();
          return { 
            success: true, 
            preferences: userData.preferences || {},
            message: 'Preferences retrieved successfully'
          };
        
        case 'set':
          if (!key || !value) {
            throw new Error('Key and value are required for set action');
          }
          
          const currentData = await user.data.get();
          const updatedData = {
            ...currentData,
            preferences: {
              ...currentData.preferences,
              [key]: value
            }
          };
          
          await user.data.update(updatedData);
          return { 
            success: true, 
            message: `Set ${key} to ${value}`,
            preferences: updatedData.preferences
          };
        
        case 'clear':
          await user.data.clear();
          return { 
            success: true, 
            message: 'All preferences cleared' 
          };
        
        default:
          throw new Error('Invalid action. Use: get, set, or clear');
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error occurred'
      };
    }
  }

  getInputSchema() {
    return z.object({
      action: z.enum(['get', 'set', 'clear']).describe('Action to perform: get (retrieve preferences), set (update a preference), or clear (remove all preferences)'),
      key: z.string().optional().describe('Preference key (required for set action)'),
      value: z.string().optional().describe('Preference value (required for set action)')
    });
  }
}
